webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultStudentGroupBatchService.java

```diff
@@ -175,35 +176,25 @@ private void archiveFile(final MultipartFile file,
 
             final Iterator<CSVRecord> recordIterator = parser.iterator();
             if (!recordIterator.hasNext()) {
-                validationFailures.add(CsvValidationResult.builder()
-                        .row(0)
-                        .message("Empty file")
-                        .build());
-                return validationFailures;
+                return newArrayList(failure(0, "Empty file"));
             }
 
             final List<String> orderedHeaders = newArrayList(parser.getHeaderMap().keySet());
             final Ordering<String> ordering = Ordering.natural().onResultOf(header -> parser.getHeaderMap().get(header));
             orderedHeaders.sort(ordering);
-            validationFailures.addAll(validationService.validateHeaders(orderedHeaders));
-            validationFailures.addAll(validationService.validateRecords(recordIterator, permissionScope));
-
-        } catch (final IOException ex) {
-            logger.warn("Problem reading uploaded file", ex);
-            validationFailures.add(CsvValidationResult.builder()
-                    .row(0)
-                    .message("Error reading file [" + ex.getMessage() + "]")
-                    .build());
-        } catch (final IllegalArgumentException ex) {
-            logger.warn("Problem parsing uploaded file", ex);
-            validationFailures.clear();
-            validationFailures.add(CsvValidationResult.builder()
-                    .row(0)
-                    .message("File does not appear to be a valid CSV with a header row")
-                    .build());
+
+            failures.addAll(validationService.validateHeaders(orderedHeaders));
+            failures.addAll(validationService.validateRecords(recordIterator, permissionScope));
+
+        } catch (final IOException exception) {
```

Comment:
What actually throws the IOException? recordIterator.next()?
Does it make sense to handle this like the IllegalArgumentException and return a collection of messages just containing this message?
Alternatively, if this is an exception that can be thrown on any row shouldn't we pass on the row number?

Reply:
@wtritch any insight on this?

Reply:
I know this exception can be thrown if there is a problem opening/reading the uploaded file.  I would assume that could happen anywhere during CSV streaming, although it's pretty unlikely after the initial load.  It's definitely unrecoverable, so I think returning at this point is ok.
I'm wondering if we actually want to change the IllegalArgumentException handling to also append to failures, rather than assume it's the one-and-only.  I'd hate to throw any other known failures on the floor when we could instead send them back to the user.

Reply:
sounds good. I'll make them both append and return the current state.